lib/mayaHydra/hydraExtensions/sceneIndex/mayaHydraMayaUsdProxyShapeSceneIndex.cpp

```diff
@@ -0,0 +1,134 @@
+    : ParentClass(sceneIndexChainLastElement)
+    , _usdImagingStageSceneIndex(usdImagingStageSceneIndex)
+    , _proxyStage(proxyStage)
+    , _dagNodeHandle(dagNodeHandle)
+{
+    TfWeakPtr<MayaUsdProxyShapeSceneIndex> ptr(this);
+    TfNotice::Register(ptr, &MayaUsdProxyShapeSceneIndex::_StageSet);
+    TfNotice::Register(ptr, &MayaUsdProxyShapeSceneIndex::_ObjectsChanged);
+    
+    _timeChangeCallbackId = MEventMessage::addEventCallback("timeChanged", onTimeChanged, this);
+}
+
+MayaUsdProxyShapeSceneIndex::~MayaUsdProxyShapeSceneIndex()
+{
+    MMessage::removeCallback(_timeChangeCallbackId);
+}
+
+MayaUsdProxyShapeSceneIndexRefPtr MayaUsdProxyShapeSceneIndex::New(const MAYAUSDAPI_NS::ProxyStage& proxyStage, 
+                                                                   const HdSceneIndexBaseRefPtr& sceneIndexChainLastElement,
+                                                                   const UsdImagingStageSceneIndexRefPtr& usdImagingStageSceneIndex,
+                                                                   const MObjectHandle& dagNodeHandle)
+{
+    return TfCreateRefPtr(new MayaUsdProxyShapeSceneIndex(proxyStage, sceneIndexChainLastElement, usdImagingStageSceneIndex, dagNodeHandle));
+}
+
+void MayaUsdProxyShapeSceneIndex::onTimeChanged(void* data)
+{
+    auto* instance = reinterpret_cast<MayaUsdProxyShapeSceneIndex*>(data);
+    if (!TF_VERIFY(instance)) {
+        return;
+    }
+    instance->UpdateTime();
+}
+
+void MayaUsdProxyShapeSceneIndex::UpdateTime()
+{
+    if (_usdImagingStageSceneIndex && _proxyStage.isValid() && _dagNodeHandle.isValid()) {
+        _usdImagingStageSceneIndex->SetTime(_proxyStage.getTime());//We have the possibility to scale and offset the time in _proxyShapeBase
+    }
+}
+
+void MayaUsdProxyShapeSceneIndex::_StageSet(const MAYAUSDAPI_NS::ProxyStageSetNotice& notice) 
+{ 
+    Populate(); 
+}
+
+void MayaUsdProxyShapeSceneIndex::_ObjectsChanged(
+    const MAYAUSDAPI_NS::ProxyStageObjectsChangedNotice& notice)
+{
+    _PopulateAndApplyPendingChanges();
+}
+
+void MayaUsdProxyShapeSceneIndex::_PopulateAndApplyPendingChanges() 
+{ 
+    Populate();
+    _usdImagingStageSceneIndex->ApplyPendingUpdates();
+}
+
+void MayaUsdProxyShapeSceneIndex::Populate()
+{
+    if (!_populated && _proxyStage.isValid()) {
+        auto stage = _proxyStage.getUsdStage();
+        // Check whether the pseudo-root has children
+        if (stage && (!stage->GetPseudoRoot().GetChildren().empty())) {
+            _usdImagingStageSceneIndex->SetStage(stage);
+            // Set the initial time
+            UpdateTime();
+            _populated = true;
+        }
+    }
+}
+
+Ufe::Path MayaUsdProxyShapeSceneIndex::InterpretRprimPath(
+    const HdSceneIndexBaseRefPtr& sceneIndex,
+    const SdfPath&                path)
+{
+    if (MayaUsdProxyShapeSceneIndexRefPtr proxyShapeSceneIndex = TfStatic_cast<MayaUsdProxyShapeSceneIndexRefPtr>(sceneIndex)) {
```

Comment:
This line is suspicious.  TfStatic_cast does not do a dynamic cast; it is therefore equivalent to doing a nullptr check on the input sceneIndex.  I wonder if a dynamic cast was intended?